docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to identify the series uniquely, queries using multiple tenants should inject a label with the name `__tenant__` and its value containing the tenant ID into the results. An potentially existing label with the same name should be overwritten.
+
+### Exposure of feature to the user
+
+#### Challenge
+
+The tenant ID is currently read from the `X-Scope-OrgID` HTTP header. There is no limitation on character classes defined, that can be used as part of the tenant ID.
+
+#### Proposal
+
+For the query path a user should be able to specify a `X-Scope-OrgID` per tenant ID and the information should be propagating as a string slice throughout out the system.
+
+To allow such queries to be processed we suggest that an experimental configuration flag `-experimental.querier.tenant-federation` will be added, which is switched off by default.
+
+### Implementing Limits and Observability for Cross-Tenant queries
+
+#### Challenge
```

Comment:
**Thought:** An additional challenge would be how we will round robin queries from the query-frontend to the queriers when multiple tenant ids exist in a single query.

Currently, we keep a queue for each tenant in the query frontend and round-robin queries from each queue to keep scheduling fair between tenants.

Reply:
I have added this in the latest commit